File: leetcode/rust/p1-two-sum.rs
// LeetCode 1. Two Sum.
use std::collections::HashMap;
use std::convert::TryFrom;
use std::vec::Vec;

struct Solution {}
impl Solution {
    pub fn two_sum(nums: &Vec<i32>, target: i32) -> Vec<i32> {
        let mut hm: HashMap<i32, i32> = HashMap::new();

        for (pos, n) in nums.iter().enumerate() {
            let ans = target - n;
            if hm.contains_key(&ans) {
                return vec![
                    hm[&ans],
                    i32::try_from(pos).expect("Bad usize -> i32 conversion"),
                ];
            }
            hm.insert(*n, i32::try_from(pos).expect("Bad usize -> i32 conversion"));
        }
        return vec![0, 0];
    }
}

fn main() {
    struct TestCase {
        nums: Vec<i32>,
        target: i32,
        exp: Vec<i32>,
    }

    let mut test_cases = Vec::<TestCase>::new();
    test_cases.push(TestCase {
        nums: vec![2, 7, 11, 15],
        target: 9,
        exp: vec![0, 1],
    });
    test_cases.push(TestCase {
        nums: vec![3, 2, 4],
        target: 6,
        exp: vec![1, 2],
    });
    test_cases.push(TestCase {
        nums: vec![3, 3],
        target: 6,
        exp: vec![0, 1],
    });

    for tc in test_cases.iter() {
        let act = Solution::two_sum(&tc.nums, tc.target);
        if tc.exp != act {
            println!(
                "FAIL. nums: {:?}, target: {}, exp: {:?}, act: {:?}",
                tc.nums, tc.target, tc.exp, act
            );
        }
    }
}

File: leetcode/py/p1-two-sum.py
from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # value -> index mapping
        d = {}
        for x, n in enumerate(nums):
            ans = target - n
            if ans in d:
                return [d[ans], x]
            d[n] = x


if __name__ == "__main__":
    s = Solution()
    tests = (
        ([1, 1], 2, [0, 1]),
        ([2, 7], 9, [0, 1]),
        ([2, 7, 11, 15], 9, [0, 1]),
        ([2, 7, 11, 15], 13, [0, 2]),
        ([2, 7, 11, 15], 17, [0, 3]),
        ([2, 7, 11, 15], 18, [1, 2]),
        ([2, 7, 11, 15], 22, [1, 3]),
        ([2, 7, 11, 15], 26, [2, 3]),
    )
    for nums, target, exp in tests:
        ans = s.twoSum(nums, target)
        if exp != ans:
            print(f"FAIL. nums: {nums}, target: {target}, exp: {exp}, ans: {ans}")
